File: bots/intgrah/v54.7.0/builder/tasks/build_foundry.py
from __future__ import annotations

from typing import TYPE_CHECKING, Final, override

from building import BuildingArmouredConveyor, BuildingConveyor
from cambc import Controller, EntityType, Environment, Position
from util.debug import debug as log

from builder.helpers import ax_feeds_target, can_afford, make_move, try_place
from builder.tasks.rejected import TaskRejectedError

if TYPE_CHECKING:
    from building import Building

    from builder import Builder

__all__ = ["build_foundry"]


FOUNDRY_ROUND_GATE: Final = 500
"""First foundry >= turn 500."""


class TaskRejectedFoundryTooEarlyError(TaskRejectedError):
    def __init__(self, current_round: int) -> None:
        self.current_round = current_round

    @override
    def __str__(self) -> str:
        return f"round {self.current_round} < gate {FOUNDRY_ROUND_GATE}"


class TaskRejectedNoFoundryTargetError(TaskRejectedError):
    @override
    def __str__(self) -> str:
        return "foundry_target is None"


class TaskRejectedFoundryTargetWrongTypeError(TaskRejectedError):
    def __init__(self, target: Position, got: Building | None) -> None:
        self.target = target
        self.got = got

    @override
    def __str__(self) -> str:
        return f"{self.target}: got {type(self.got).__name__}, expected Ti conveyor"


class TaskRejectedFoundryTargetEnemyError(TaskRejectedError):
    def __init__(self, target: Position) -> None:
        self.target = target

    @override
    def __str__(self) -> str:
        return f"{self.target}: conveyor held by enemy team"


class TaskRejectedFoundryTerrainNotEmptyError(TaskRejectedError):
    def __init__(self, target: Position, env: Environment | None) -> None:
        self.target = target
        self.env = env

    @override
    def __str__(self) -> str:
        return f"{self.target}: terrain is {self.env}, not EMPTY"


class TaskRejectedAxChainNotConnectedError(TaskRejectedError):
    def __init__(self, target: Position) -> None:
        self.target = target

    @override
    def __str__(self) -> str:
        return f"ax chain hasn't reached {self.target}"


def build_foundry(self: Builder, ct: Controller) -> None:
    """Place a foundry at `self.foundry_target` once an Ax-sourced chain
    feeds into it. If affordability is the only blocker, the builder walks
    to the target and waits rather than wandering off (foundries cost more
    than a harvester's worth of Ti and that buffer must not be spent on
    other work while the chain sits idle)."""
    if self.round < FOUNDRY_ROUND_GATE:
        raise TaskRejectedFoundryTooEarlyError(self.round)
    target = self.foundry_target
    if target is None:
        raise TaskRejectedNoFoundryTargetError
    bld = self.get_building(target)
    if not isinstance(bld, BuildingConveyor | BuildingArmouredConveyor):
        raise TaskRejectedFoundryTargetWrongTypeError(target, bld)
    if bld.team != self.my_team:
        raise TaskRejectedFoundryTargetEnemyError(target)
    if self.get_env(target) != Environment.EMPTY:
        raise TaskRejectedFoundryTerrainNotEmptyError(target, self.get_env(target))
    if not ax_feeds_target(self, target):
        raise TaskRejectedAxChainNotConnectedError(target)

    # Chain connected, tile valid. Committed — walk/wait to place.
    dist_sq = self.my_pos.distance_squared(target)
    if not can_afford(self, EntityType.FOUNDRY):
        if dist_sq <= 2:
            log(f"build_foundry: holding {target} until affordable")
            return
        log(f"build_foundry: walking toward {target}, can't afford yet")
        make_move(self, ct, target)
        return

    if self.my_pos == target:
        for d, _npos in self.dir_neighbours_4:
            if ct.can_move(d):
                ct.move(d)
                break
        else:
            log(f"build_foundry: stuck on {target}, cannot step off this turn")
            return

    if target in self.all_bots and self.all_bots[target] != self.my_id:
        log(f"build_foundry: {target} occupied by friendly bot, holding")
        return

    if self.my_pos.distance_squared(target) <= 2 and try_place(
        self,
        ct,
        EntityType.FOUNDRY,
        target,
    ):
        log(f"build_foundry: PLACED at {target}")
        return

    log(f"build_foundry: out of range of {target}, walking")
    make_move(self, ct, target)

File: bots/beacon/v670/src/builder/tasks/econ/infrastructure/build_foundry.rs
//! Translation of `bots/intgrah/v54.7.9/builder/tasks/econ/infrastructure/build_foundry.py`.
//!
//! Replace a designated Ti conveyor (`foundry_target`) with a foundry
//! once its Ax feed is established. Gated on round >= `FOUNDRY_ROUND_GATE`.
//! Checks the target is still a friendly pure conveyor, that an Ax cardinal
//! feeds it, and that we can afford the build; walks adjacent and destroys-
//! then-builds the foundry.

use cambc::{BuildExtra, Controller, ControllerApi, EntityType, Environment};

use crate::builder::Builder;
use crate::builder::helpers::{ax_feeds_target, can_afford, make_move, try_place};
use crate::builder::tasks::rejected::{TaskRejected, TaskResult};
use crate::util::debug::debug as log;
use crate::util::posint::{dist_sq, idx_of};
use serde_json::Map;

/// First foundry >= turn 500.
const FOUNDRY_ROUND_GATE: i32 = 500;

pub fn build_foundry(self_: &mut Builder, ct: &mut Controller<'_>) -> TaskResult {
    if self_.round < FOUNDRY_ROUND_GATE {
        return Some(TaskRejected::from_string(format!(
            "round {} < gate {}",
            self_.round, FOUNDRY_ROUND_GATE
        )));
    }
    let Some(target) = self_.foundry_target else {
        return Some(TaskRejected::new("foundry_target is None"));
    };
    let kind = self_.kind_at(target);
    let team = self_.team_at(target);
    let is_conveyor = matches!(
        kind,
        Some(EntityType::Conveyor | EntityType::ArmouredConveyor)
    );
    if !is_conveyor {
        return Some(TaskRejected::from_string(format!(
            "{target:?}: got {kind:?}, expected Ti conveyor"
        )));
    }
    if team != Some(self_.my_team) {
        return Some(TaskRejected::from_string(format!(
            "{target:?}: conveyor held by enemy team"
        )));
    }
    if self_.get_env(target) != Some(Environment::Empty) {
        return Some(TaskRejected::from_string(format!(
            "{:?}: terrain is {:?}, not EMPTY",
            target,
            self_.get_env(target)
        )));
    }
    // `update_foundry_target` accepts junctions (multi-input tiles with both
    // Ti and Ax history) even when the structural Ax chain hasn't reached
    // yet — e.g. a feeder is in the process of being built. Mirror that
    // tolerance here, otherwise the task rejects forever on valid junction
    // targets and the foundry is never placed.
    let target_p = idx_of(target);
    if !ax_feeds_target(self_, target) && !pyrust::vec::contains!(self_.junctions, &target_p) {
        return Some(TaskRejected::from_string(format!(
            "ax chain hasn't reached {target:?}"
        )));
    }

    let my_pos_p = idx_of(self_.my_pos);
    let d_sq = dist_sq(my_pos_p, target_p);
    if !can_afford(self_, EntityType::Foundry) {
        if d_sq <= 2 {
            log(
                &format!("build_foundry: holding {target:?} until affordable"),
                Map::new(),
            );
            return None;
        }
        log(
            &format!("build_foundry: walking toward {target:?}, can't afford yet"),
            Map::new(),
        );
        make_move(self_, ct, target);
        return None;
    }

    if self_.my_pos == target {
        let dirs = pyrust::clone!(self_.dir_neighbours_4);
        let mut moved = false;
        for (d, _npos) in dirs {
            if pyrust::unwrap!(ct.can_move(d)) {
                pyrust::unwrap!(ct.move_(d));
                moved = true;
                break;
            }
        }
        if !moved {
            log(
                &format!("build_foundry: stuck on {target:?}, cannot step off this turn"),
                Map::new(),
            );
            return None;
        }
    }

    if let Some(&uid) = self_.all_bots.get(&target)
        && uid != self_.my_id
    {
        log(
            &format!("build_foundry: {target:?} occupied by friendly bot, holding"),
            Map::new(),
        );
        return None;
    }

    if dist_sq(my_pos_p, target_p) <= 2
        && try_place(
            self_,
            ct,
            EntityType::Foundry,
            target,
            BuildExtra::None,
            true,
        )
    {
        log(&format!("build_foundry: PLACED at {target:?}"), Map::new());
        return None;
    }

    log(
        &format!("build_foundry: out of range of {target:?}, walking"),
        Map::new(),
    );
    make_move(self_, ct, target);
    None
}
